### PR TITLE
[Docs] Add documentation and examples for `Figure`

### PR DESCRIPTION
**Pull request by huong-li-nguyen**
## Description
- Add How-to-guide for `Figure`
- Add How-to-guide for custom `Figure`
- Update features examples dashboard with `Figure` and custom `Figure`
- Tidy API reference

https://vizro--524.org.readthedocs.build/en/524/pages/user-guides/figure/
https://vizro--524.org.readthedocs.build/en/524/pages/user-guides/custom-figures/

More cross-links to the KPI card functions can be found in this PR: https://github.com/mckinsey/vizro/pull/529

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-core/changelog.d/20240607_125345_huong_li_nguyen_create_kpi_cards.md
Comment (huong-li-nguyen) [inline vizro-core/changelog.d/20240607_125345_huong_li_nguyen_create_kpi_cards.md:10]: This will work after merge

File: vizro-core/changelog.d/20240607_125345_huong_li_nguyen_create_kpi_cards.md
Comment (huong-li-nguyen) [inline vizro-core/changelog.d/20240607_125345_huong_li_nguyen_create_kpi_cards.md:11]: Will be updated in a separate PR on KPI Cards

File: vizro-core/changelog.d/20240607_125345_huong_li_nguyen_create_kpi_cards.md
Comment (huong-li-nguyen) [inline vizro-core/changelog.d/20240607_125345_huong_li_nguyen_create_kpi_cards.md:10]: Link will be updated to https://vizro.readthedocs.io/en/stable/pages/user-guides/figure/ after approval, otherwise the docs link won't get generated because it's failing the build as the link is not available yet

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:3]: If you've used "custom figures" here, why capitalise it above in the title?

I think it needs to be consistent, but am not sure which you'd want to do -- probably lower case throughout? See above suggestion if so, and accept.

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:1]:
```suggestion
# How to create custom figures
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:8]:
```suggestion
* when you need a figure that isn't available in our pre-defined figure functions [`vizro.figures`](../API-reference/figure-callables.md) or
* when you need a figure that doesn't fit into our predefined components like [`Graph`][vizro.models.Graph], [`Table`][vizro.models.Table] or [`AgGrid`][vizro.models.AgGrid].
```

Comment (stichbury) [reply vizro-core/docs/pages/user-guides/custom-figures.md:8]: Not a sequence so doesn't need numbered bullets.

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:10]:
```suggestion
Use the following approach, which is similar to creation of a [custom chart](../user-guides/custom-charts.md), do the following:
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:12]:
```suggestion
1. Define a function that returns a `Dash` component.
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:13]:
```suggestion
2. Decorate it with the `@capture("figure")` respectively.
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:14]:
```suggestion
3. The function must accept a `data_frame` argument (of type `pandas.DataFrame`).
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:15]:
```suggestion
4. The figure should be derived from and require only one `pandas.DataFrame`. Additional dataframes from other arguments will not react to dashboard components such as `Filter`.
```

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:3]:
```suggestion
This guide shows you how to add any Dash component to Vizro that needs to be reactive to controls or actions in your page.
```

Comment (stichbury) [reply vizro-core/docs/pages/user-guides/figure.md:3]: Capitalised?

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:7]:
```suggestion
components in Vizro. The [`Graph`][vizro.models.Graph], [`Table`][vizro.models.Table] and [`AgGrid`][vizro.models.AgGrid] components
```

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:8]:
```suggestion
are specific implementations of `Figure`. They serve as intuitive shortcuts, embedding behaviors and
```

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:13]: We had some problems with Mermaid diagrams in Kedro whereby they stopped rendering for no real reason, so let's keep an eye on this. What we did was to keep the Mermaid in the file but comment it out and instead screengrab the image (before commenting it out). So it's available to update relatively easily but we aren't expecting it to render on the fly.

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:15]:
```suggestion
  first["`Does the Dash component you want to add exist in our components library already?`"]
```

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:16]:
```suggestion
  extend-component([Use an existing component and enhance it if required])
```

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:17]:
```suggestion
  second["`Does the new Dash component need to be reactive to controls or actions?`"]
```

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:34]:
```suggestion
To add a `Figure` to your page:
```

File: vizro-core/docs/pages/user-guides/figure.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/figure.md:37]: These should be numbered bullets and start with capitals too.

**Review by stichbury — APPROVED** [review]
🌟 Looks great to me, just a few minor change suggestions for consistency 🌟

File: vizro-core/docs/pages/user-guides/figure.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/figure.md:13]: Ah, that's good to know! We currently also use it for the dynamic data section, but currently, I haven't encountered any issues yet 👍 What I like about it is the interactivity as you can click directly to the relevant links, which wouldn't work in a screenshot

**Comment by antonymilne** [comment]
FYI @stichbury @huong-li-nguyen I pushed [ea5cf95](https://github.com/mckinsey/vizro/pull/524/commits/ea5cf959d3ae17584fe1e95e1bc7a37c3e0fbfee) so that the linkchecker ignores links to our built RTD - otherwise it's impossible to add new pages to our docs since it will always fail on CI like you found here. 

I've also pushed https://github.com/mckinsey/vizro/pull/534 which will make our checks stricter so that it's not possible to link to a page that doesn't exist in our docs. This means that there's no need to check the built RTD docs any more.

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:59]: This doesn't match the type hint! `vm.Card.build` returns `dbc.Card`.

When I was writing this example code I thought about two ways of doing it:
1. do `vm.Card.build`: means you don't need to understand any more Dash, just Vizro, but it uses the `build` method which maybe looks weird and means creating many new models on the fly
2. do `dbc.Card` directly: means you need to understand Dash but doesn't add the complications of Vizro models

In the case of answering the user I chose to go for 1, but here I think we should probably go for option 2. Otherwise it's not clear how you can use a custom figure to inject an arbitrary dash component - it just looks like you are still restricted to vizro models.

We should also consider whether we want to encourage people to return a list of Dash components or whether they should actually be forced to return one. Probably we should just restrict to only one like we do everywhere else (even though a list would probably still work fine because it just gets put in a html.Div or dcc.Loading or whatever afterewards).

So I'd suggest something like this:
```suggestion
            return html.Div(dbc.Card(dcc.Markdown(f"### Card #{i+1}\n{text}")) for i, text in enumerate(texts))
            # or if we think allowing list is ok
            return [dbc.Card(dcc.Markdown(f"### Card #{i+1}\n{text}")) for i, text in enumerate(texts)]
```
and change the type hint accordingly.

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:69]: This could do with a `(4)!` to explain the `n_rows` thing and link to the page on parameters.

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:21]: I really like this example but it's quite advanced because it's doing several complicated things:
* setup of mock data (if you can find a better way to do this using some built in `px.data` dataset then please do so)
* custom CSS
* function returns multiple Dash components
* parameter to control a function

Painful though it may be to write 😅 I think we should have a total of three examples on this page which increase in complexity and show different things:
1. a custom KPI card that basically copies and pastes code from our API reference docs and then modifies a simple thing, like you suggest in https://github.com/mckinsey/vizro/pull/529/files#r1644213316
2. a really simple example that writes a whole new non-KPI card Dash component that reacts to a `Filter` (lmk if you need help thinking of some example) 
3. this multiple cards example

File: vizro-core/mkdocs.yml
Comment (antonymilne) [inline vizro-core/mkdocs.yml:50]: The custom figure page doesn't appear on the navbar at the moment so I think needs to be added here?

**Review by antonymilne — COMMENTED** [review]
Really great stuff ⭐ The API docs are top-notch also, and glancing at https://github.com/mckinsey/vizro/pull/529/ the additions to docs there look great too - I'll review properly later.

Please can we add reciprocal links with the Cards & buttons page once you've added the KPI stuff in #529? Just to clarify the relationship between a card and a KPI card.

I've left lots of comments but leave it to you to decide exactly how many you agree with/want to do here vs. in a future PR 🙂 Since I know writing docs and examples for this is a huge job and I don't want it to become too painful for you!

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/custom-figures.md:60]: Ah, yes - that makes sense!

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/custom-figures.md:22]: Alright, sounds good. On two, we could do a simple html.H3 component that dynamically changes the title based on a filter. I think this would the simplest example, especially if we don't want to have another card example 😄

Comment (antonymilne) [reply vizro-core/docs/pages/user-guides/custom-figures.md:22]: Yeah, I think the challenge is to come up with something that really requires `data_frame` and doesn't require a `dbc.Card` 🤔 (Maybe in the future we will make this no longer a requirement since you might like a component that reacts to a parameter and isn't connected to a dataframe, but let's leave it as required for now anyway.)

e.g. if we show something like `len(data_frame)` then it's probably best shown just in a KPI (or non-KPI) card. So if we did something like `return html.H3(f"Your data has {len(data_frame)} rows")` then maybe it will always look kind of stupid compared to just putting it in a `dbc.Card`. 

So maybe my idea of `Filter` + non-card + `data_frame` required is actually stupid here and will just result in a really weird convoluted example 😅

Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/custom-figures.md:22]: I don't think your idea is _that_ stupid 😄 To have a title that dynamically changes based on a filter/parameter is not that uncommon. I can think of a use-case e.g. where you have a business dashboard and you want to dynamically change the title to reflect the department you are filtering on e.g. "Summary for Department {XXX}".

It does make more sense with other components on the page, but I think adding that simple example with an html.Div might be useful already, just to show that you can return any Dash component not just a dbc.Card 😄

Comment (antonymilne) [reply vizro-core/docs/pages/user-guides/custom-figures.md:22]: Cool, sounds good 😀 The problem with updating text based on the filter value is that the value itself isn't actually available directly, you'd have to do `data_frame[column].something`, which also feels a bit awkward and would maybe be more cleanly done through an action. But probably it's fine just as an example to demonstrate the sort of thing you can do.

**Comment by huong-li-nguyen** [comment]
@antonymilne @stichbury - could the two of you take a final look? Thank you! 🙏

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:15]:
```suggestion
1. Define a function that returns a [Dash component](https://dash.plotly.com/#open-source-component-libraries). This can, but does not need to, be based on code in our pre-defined figure functions in [`vizro.figures`](../API-reference/figure-callables.md).
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:30]:
```suggestion
For instance, to make a KPI card with the icon positioned on the right side of the title instead of the left, copy and paste the [source code of `kpi_card`](API-reference/figure-callables/#vizro.figures.kpi_card) and
adjust the return statement of the function.
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:109]:
```suggestion
        3. We adjust the return statement to include the icon on the right side of the title. This is achieved by swapping the order of the `html.H2` and `html.P` compared to the original `kpi_card`.
```
(I think that's the only change?)

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:162]:
```suggestion
        4. Our custom figure function `dynamic_html_header` now needs to be passed on to the `vm.Figure`.
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:222]:
```suggestion
                [dbc.Card(dcc.Markdown(f"### Card #{i}\n{text}")) for i, text in enumerate(texts, 1)],
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/custom-figures.md:203]: This is super ugly, even if it's gone through black. 😱 I would un-black it or rewrite the dataframe somehow to make it easier to see, like just break it down:
```
text = [
   "blah", 
   "blah"
]
df = pd.DataFrame({"text": text * 2})
```

**Review by antonymilne — APPROVED** [review]
Really great set of examples! Thank you for all your hard work on this 🙏

File: vizro-core/docs/pages/API-reference/figure-callables.md
Comment (stichbury) [inline vizro-core/docs/pages/API-reference/figure-callables.md:4]:
```suggestion
[`Figure`][vizro.models.Figure]. The [how-to guide on figures](../user-guides/figure.md)
```

File: vizro-core/docs/pages/API-reference/figure-callables.md
Comment (stichbury) [inline vizro-core/docs/pages/API-reference/figure-callables.md:5]:
```suggestion
contains more information.
```

File: vizro-core/docs/pages/user-guides/custom-charts.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-charts.md:7]:
```suggestion
In general, you should use the custom chart decorator `@capture("graph")` if your plotly chart needs any post-update calls or customization e.g:
```

Comment (stichbury) [reply vizro-core/docs/pages/user-guides/custom-charts.md:7]: Is Vale running OK? I would have expected an error for the e.g. here. When you've successfully merged this PR, next sprint when I am back on Vizro tickets, I'll take a look to find out why it didn't.

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/custom-figures.md:21]:
```suggestion
The following examples can be used as a base to build more sophisticated figures.
```

File: vizro-core/docs/pages/user-guides/custom-figures.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/custom-figures.md:6]: Ah okay, I took it over from the custom-charts.md -  we should probably consolidate these at some point 😬

File: vizro-core/docs/pages/user-guides/custom-charts.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/custom-charts.md:7]: Interesting - yes, there are no errors raised or at least it's passing CI!

**Review by stichbury — APPROVED** [review]
LGTM 🌟

**Pull request merged**
